Thanks — all four applied in 73a5cd0.

**1. Test flakiness.** Agreed, and the asymmetry you point at is the real problem: the two windows are disjoint *and* unequal, so a burst landing in the short one has nothing to move the floor with. `got` is now the best of three `F` windows. An unbounded loop stalls every window, so the signal survives; one unlucky burst no longer fails the run. On n0 with AVX2: `F 375.306µs (best of 3), pure-Go fGeneric 1.641865ms`.

**2. Compile-time assert.** Added — nice trick, and it earns its place for exactly the reason you give: on non-AVX2 hardware `TestFChunkedMatchesGeneric` compares `fGeneric` against itself, so nothing else catches a bad retune. Verified in both directions — 4090 builds, and 4096 fails with:

```
blake2b.go:110:16: cannot use -(maxAsmRounds % 10) (untyped int constant -6) as uint value in constant declaration (overflows)
```

**3. `fLong` fallback comment.** Added. Worth being explicit that it is not just a perf choice: `fAVX` and `fSSE4` derive the working vector from `h` and fold it back inside one call, so they cannot be handed a chunk at all — routing that branch through `f()` would silently reintroduce the unbounded loop on pre-Haswell hardware.

**4. Skip guards.** Split into a `testing.Short()` guard and a separate `GOMAXPROCS >= 2` capability skip.

---

On the Copilot comment above about the cumulative histogram: it is correct in general — that metric is cumulative for the process lifetime and reading a max over it does attribute earlier pauses to the current code. It is a trap I hit while developing this. But it applies to `blake2b_preempt_test.go`, an earlier revision of this branch that no longer exists; the current `blake2b_f_test.go` uses `runtime.GC()` wall time via `gcWait` and reads no histogram at all (`grep -c 'sched/pauses' == 0`). So nothing to change there.

--- a/common/crypto/blake2b/blake2b.go
+++ b/common/crypto/blake2b/blake2b.go
@@ -96,18 +96,17 @@ func New256(key []byte) (hash.Hash, error) { return newDigest(Size256, key) }
 // and BinaryUnmarshaler for state (de)serialization as documented by hash.Hash.
 func New(size int, key []byte) (hash.Hash, error) { return newDigest(size, key) }
 
+// sigmaRounds is the length of the BLAKE2b message-schedule cycle: round i
+// permutes with precomputed[i%sigmaRounds]. The assembly unrolls a whole cycle
+// and loops over it, so a chunk boundary is only correct on a multiple of it.
+const sigmaRounds = 10
+
 // maxAsmRounds bounds the work a single assembly call does. Assembly is not
 // preemptible, so an unbounded rounds argument holds every P in stop-the-world
-// for the length of the call. It must stay a multiple of 10 so that the message
-// schedule realigns at a chunk boundary.
+// for the length of the call.
 const maxAsmRounds = 4090
 
-// The assembly unrolls the ten distinct message permutations of the round
-// function and loops over that block, so a chunk boundary is only correct on a
-// multiple of ten. Enforced at compile time: the expression is 0 when the
-// remainder is 0, and a negative untyped constant otherwise, which cannot
-// convert to uint. Builds at 4090, fails at 4096.
-const _ uint = -(maxAsmRounds % 10)
+const _ uint = -(maxAsmRounds % sigmaRounds) // compile-time check: multiple of sigmaRounds
 
 // F is a compression function for BLAKE2b. It takes as an argument the state
 // vector `h`, message block vector `m`, offset counter `t`, final block indicator

--- a/common/crypto/blake2b/blake2b_generic.go
+++ b/common/crypto/blake2b/blake2b_generic.go
@@ -9,9 +9,9 @@ import (
 )
 
 // the precomputed values for BLAKE2b
-// there are 10 16-byte arrays - one for each round
+// there is one 16-byte array per round of the message-schedule cycle
 // the entries are calculated from the sigma constants.
-var precomputed = [10][16]byte{
+var precomputed = [sigmaRounds][16]byte{
 	{0, 2, 4, 6, 1, 3, 5, 7, 8, 10, 12, 14, 9, 11, 13, 15},
 	{14, 4, 9, 13, 10, 8, 15, 6, 1, 0, 11, 5, 12, 2, 7, 3},
 	{11, 12, 5, 15, 8, 0, 2, 13, 10, 3, 7, 9, 14, 6, 1, 4},
@@ -32,7 +32,7 @@ func fGeneric(h *[8]uint64, m *[16]uint64, c0, c1 uint64, flag uint64, rounds ui
 	v14 ^= flag
 
 	for i := 0; i < int(rounds); i++ {
-		s := &(precomputed[i%10])
+		s := &(precomputed[i%sigmaRounds])
 
 		v0 += m[s[0]]
 		v0 += v4

